src/tls_relation.py

```diff
@@ -5,35 +5,54 @@
 """Gateway API TLS relation business logic."""
 import secrets
 import string
-from typing import Dict, Union
+import typing
 
 from charms.tls_certificates_interface.v3.tls_certificates import (
-    CertificateAvailableEvent,
     CertificateExpiringEvent,
     CertificateInvalidatedEvent,
+    ProviderCertificate,
     TLSCertificatesRequiresV3,
     generate_csr,
     generate_private_key,
 )
 from cryptography import x509
 from cryptography.x509.oid import NameOID
-from ops.jujuversion import JujuVersion
-from ops.model import Model, Relation, SecretNotFoundError
+from ops.model import Relation, SecretNotFoundError
 
 TLS_CERT = "certificates"
 
 
+def get_hostname_from_cert(certificate: str) -> str:
+    """Get the hostname from a certificate subject name.
+
+    Args:
+        certificate: The certificate in PEM format.
+
+    Returns:
+        The hostname the certificate is issue to.
+    """
+    decoded_cert = x509.load_pem_x509_certificate(certificate.encode())
+
+    common_name_attribute = decoded_cert.subject.get_attributes_for_oid(NameOID.COMMON_NAME)
+    if not common_name_attribute:
+        return ""
```

Comment:
Not sure we should return an empty string.
If you think we should, then when you call that method you should be defensive

Reply:
I think the idea behind this was that we can trust that the provider will send us valid certificates so we don't need to be defensive. I've thought about it for a while and I agree that doing it like this is a bit fragile since now the method is used in way more places than in the `nginx-ingress-integrator` charm and the output of this is used to lookup secrets and re-requesting new certs as well so raising an exception here makes sense for me. I've add that code in and handle the thrown exception in `validation.py`